libraries/AP_RangeFinder/AP_RangeFinder_TeraRangerDUOSerial.cpp

```diff
@@ -0,0 +1,147 @@
+{
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+// detect if a TeraRanger DUO rangefinder sensor is connected by looking for a configured serial port
+bool AP_RangeFinder_TeraRangerDUOSerial::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// update the state of the sensor
+void AP_RangeFinder_TeraRangerDUOSerial::update(void)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    // process incoming messages
+    if (get_reading(state.distance_cm)) {
+        _last_reading_ms = AP_HAL::millis();        
+    }
+    update_status();
+}
+
+// update status
+void AP_RangeFinder_TeraRangerDUOSerial::update_status()
+{
+    if (AP_HAL::millis() - _last_reading_ms > TRDUO_TIMEOUT_MS) {
+        set_status(RangeFinder::RangeFinder_NoData);
+    } else {
+        if ((int16_t)state.distance_cm > TRDUO_MAX_DISTANCE_TOF) {
+            set_status(RangeFinder::RangeFinder_OutOfRangeHigh);
+        } else if ((int16_t)state.distance_cm < TRDUO_MIN_DISTANCE_SOUND) {
+            set_status(RangeFinder::RangeFinder_OutOfRangeLow);
+        } else {
+            set_status(RangeFinder::RangeFinder_Good);
+        }
+    }
+}
+
+// read - return last value measured by sensor
+bool AP_RangeFinder_TeraRangerDUOSerial::get_reading(uint16_t &reading_cm)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_tof = 0;
+    float sum_sound = 0;
+    uint16_t tof_count = 0;
+    uint16_t sound_count = 0;
+    int16_t nbytes = uart->available();
+
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        if (c == 'T') {
+            _buffer_count = 0;
+            _found_start = true;
+        }
+
+        if (!_found_start) {
+            continue;
+        }
+
+        _buffer[_buffer_count++] = c;
+        
+        /*
+            The distance information is a message of exactly 7 bytes, the headers
+            are the character T(0x54) at byte one and S (0x53) at byte four.
+            Each header is followed by two bytes for the distance value in millimeter.
+            Byte seven is the checksum(CRC8) of all 6 bytes before.
+        */
+        if (_buffer_count >= TRDUO_BUFFER_SIZE_FULL) {
+            // check if message has right CRC
+            if (crc_crc8(_buffer, TRDUO_BUFFER_SIZE_FULL - 1) == _buffer[TRDUO_BUFFER_SIZE_FULL - 1]){
+                uint16_t t_distance = UINT16_VALUE(_buffer[1], _buffer[2]);
+                uint16_t s_distance = UINT16_VALUE(_buffer[4], _buffer[5]);
+
+                // check ToF distance range
+                if (t_distance > TRDUO_MIN_DISTANCE_TOF && t_distance < TRDUO_MAX_DISTANCE_TOF) {
```

Comment:
14m (in non-direct sunlight conditions, ≥ 6m otherwise) for ToF,Sonar up to 7.65m
1kHz (in speed mode) - up to 600Hz (in precision mode), Sonar 1-20Hz

The combination of the TeraRanger One's infrared time-of-flight technology and high quality sonar allows the sensor to detect from transparent, reflective, bright or even acoustical surfaces.
I think that it is better to use the distance of the ultrasonic wave when it is 7.65m or less.
I think that ultrasound is effective in the following cases.
In addition, I think that I can detect a window glass when using it for an auction.

= □ = Drone



ー ー ー ー Glass




++++++ ground